### ISSUE TITLE
Add pub (Dart) as supported package manager

### ISSUE THREAD

**Issue by tsteenbe** [heremaps/oss-review-toolkit#757]
Add support for the pub dependency manager to the analyzer module.

https://github.com/dart-lang/pub

Possible test project using pub: https://github.com/sass/dart-sass

**Comment by moo24** [comment]
### Key question: Is it possible in Dart to get a dependency tree that includes package names, version, scopes licenses and source locations?

#### the package dependency tree can be generated using
```sh
pub deps --style=list
```
#### which provides the following output (where NameOfPackage is the name of the package this command is run for):
```sh
Dart SDK 2.2.1-dev.4.0.flutter-None
Flutter SDK 1.4.9-hotfix.1
NameOfPackage 0.8.0

dependencies:
- protobuf 0.13.11
  - fixnum ^0.10.9
- flutter 0.0.0
  - sky_engine any
  - vector_math 2.0.8
  - collection 1.14.11
  - typed_data 1.1.6
  - meta 1.1.6

dev dependencies:
- test 1.6.1
  - async >=1.13.0 <3.0.0
  - shelf_static ^0.2.6
  - analyzer >=0.26.4 <0.37.0

transitive dependencies:
- args 1.5.1
- async 2.1.0
  - collection ^1.5.0
- yaml 2.1.15
  - charcode ^1.1.0
  - source_span >=1.0.0 <2.0.0
  - string_scanner >=0.1.4 <2.0.0
  - collection
```
(note: I truncated the output).

#### Another option is to look into the pubspec.lock (https://dart.dev/tools/pub/glossary#lockfile) file. 
Unlike the pubspec, which only lists immediate dependencies and allows version ranges, the lock file comprehensively pins down the entire dependency graph to specific versions of packages. 
A lockfile ensures that you can recreate the exact configuration of packages used by an application.
The package hosting location is also stored in the pubspec.lock file.
```sh
cat pubspec.lock 
```
```yaml
Generated by pub
See https://www.dartlang.org/tools/pub/glossary#lockfile
packages:
  async:
    dependency: transitive
    description:
      name: async
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.0.8"
  dart_config:
    dependency: transitive
    description:
      path: "."
      ref: HEAD
      resolved-ref: e26916e095244a7e5ea61315b030d298d127ed26
      url: "https://github.com/MarkOSullivan94/dart_config.git"
    source: git
    version: "0.5.0"
  dart_style:
    dependency: transitive
    description:
      name: dart_style
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.2.2"
```
(note: I truncated the output)


#### packages are installed to 
~/.pub-cache/hosted/pub.dartlang.org/

e.g. the package contents 'async' with version '2.2.0' will be stored in
~/.pub-cache/hosted/pub.dartlang.org/async-2.2.0

#### licenses are part of package and the file name is always LICENSE. As far as I know, a LICENSE file is a mandatory requirement to publish packages on pub.dartlang.org.
```sh
cat ~/.pub-cache/hosted/pub.dartlang.org/async-2.2.0/LICENSE
```

### One more thing to consider for supporting Flutter packages

A Dart package which is a flutter plugins can contain native iOS and/or Android code. Flutter is currently using Cocoapods on iOS and Gradle for Android to install the native dependencies for each platform. While Gradle is already support, cocoapods are not yet supported.

**Comment by sschuberth** [comment]
Done in #1539.
